### ISSUE TITLE
utils/grid2pdf broken/abandoned?

### ISSUE THREAD

**Issue by SJaffa** [danieljprice/splash#46]
Hi Daniel,

I'm trying to make column density PDFs of a bunch of simulations. Since Splash can produce the column density grids for me from my Gandalf dumpfiles I was hoping there would be a way to produce the PDFs as well, as I don't need the column density grids for much else. I found the [grid2pdf](https://github.com/danieljprice/splash/blob/master/utils/grid2pdf.f90) script in the utils folder which looks like it might be what I want but it doesn't seem to be referenced anywhere else in the code or actually used. Is it meant to be used as a standalone script? Or is it a relic that is no longer implemented (I see it was edited 10 years ago)?

You also mention in the [documentation](https://splash-viz.readthedocs.io/en/latest/other.html#using-splash-to-time-average-a-series-of-files) that the 'calc timeaverage' could be used to get "time averaged probability density function (PDF) from PDFs produced by splash ." but again no mention of how to produce the individual PDFs.

Thanks,
Sarah

**Comment by danieljprice** [comment]
Hi Sarah,

Thanks for spotting, looks like I broke the build on this accidentally when cleaning up the pdf-generation module a while ago - now fixed.

From looking at the code, you will need to output grids in "binary" format, i.e.

splash to gridbinary turb_00020

or if you want to skip the velocity interpolation (assuming density in column 6):

SPLASH_TO_GRID=6 splash to gridbinary turb_00020

this produces a file called turb_00020.grid, then follow this with:

cd splash; make grid2pdf
cd -
~/splash/bin/grid2pdf turb_00020.grid

which produces:

turb_00020.grid_pdf_ln_density.dat

this is just a two-column ascii file, so you can then plot this with your favourite plotting tool, e.g.

splash -ev turb_00020.grid_pdf_ln_density.dat

Best  wishes,

Daniel

**Comment by SJaffa** [comment]
Great, thanks for your help.

**Comment by SJaffa** [comment]
Hi Daniel,

Sorry I thought this was working but I've run into another problem. Firstly, the limits of the PDF (rhologmin, rhologmax) are hardcoded which seems like a bad idea. I've changed my local copy to find the minimum and maximum of the rhogrid it read from the binary file and use these as the limits. However, it seems to be messing up somewhere. 

When I create the grid file (srsplash to gridbinary dumpfile) it reports the min and max density on the grid as around 1e-38 to 1e-20. The same is try if I use srsplash to grid and make an ascii file. But when I then use grid2pdf dumpfile.grid it reports the min and max density as 1e-306, 1e-136 (not logged, I got it to find and print the min and max before logging the rhogrid.)

**Comment by SJaffa** [comment]
Hi Daniel,

Sorry I thought this was working but I've run into another problem. Firstly, the limits of the PDF (rhologmin, rhologmax) are hardcoded which seems like a bad idea. I've changed my local copy to find the minimum and maximum of the rhogrid it read from the binary file and use these as the limits. However, it seems to be messing up somewhere. 

When I create the grid file (srsplash to gridbinary dumpfile) it reports the min and max density on the grid as around 1e-38 to 1e-20. The same is try if I use srsplash to grid and make an ascii file. But when I then use grid2pdf dumpfile.grid it reports the min and max density as 1e-306, 1e-136 (not logged, I got it to find and print the min and max before logging the rhogrid.)

**Comment by danieljprice** [comment]
I guess the point is that -15 to 15 is a large range (30 orders of magnitude!) so I figured it would cover most use cases. The issue is that if you are time averaging a bunch of PDFs then you want the bins to be the same, so I did not want to dynamically adjust the min and max for the logrho grid.

Using splash to grid tries to avoid giving you empty cells by finding the minimum density on the SPH particles and setting this as a floor on the density on the grid (which can be adjusted with SPLASH_TO_GRID_RHOMIN=0.0), so this could be the reason for the different minima being reported?
